clang/lib/Format/TokenAnnotator.cpp

```diff
@@ -4591,7 +4589,9 @@ bool TokenAnnotator::spaceRequiredBetween(const AnnotatedLine &Line,
     if (!BeforeLeft)
       return false;
     if (BeforeLeft->is(tok::coloncolon)) {
-      return Left.is(tok::star) &&
+      const auto *Prev = BeforeLeft->Previous;
+      return Left.is(tok::star) && Prev &&
+             !Prev->endsSequence(tok::identifier, TT_FunctionTypeLParen) &&
              Style.PointerAlignment != FormatStyle::PAS_Right;
```

Comment:
nit: i think `PointerAlignment` is always `non-right` when we get here, as we bail out on line `4577` when pointer alignment is right (or it's derived to be right).

Reply:
Good catch!